src/nupic/research/boosting.py

```diff
@@ -0,0 +1,387 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2013-2014, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+
+from nupic.bindings.math import (GetNTAReal,
+                                 Random as NupicRandom)
+
+realDType = GetNTAReal()
+
+
+
+class Boosting(object):
+  """
+  Implements boosting for columns of the spatial pooler.
+  Boosting is an optional process that helps the spatial pooler (SP) to adapt
+  faster to changes in the input stream. Boosting acts on 2 levels: "duplicit
+  peer columns activity" and "mismatched input overlap". 
+
+  1/ Boosting of mispatched input:
+  Triggered when the column is "growing" (connected to) the inputs that are 
+  not used (anymore). This can happen in two cases, when a part of the input
+  vector is never used (wrongly, too generic designed encoder, etc.), or when
+  a part of the input that had been used before is not useful anymore for a long
+  period of time. 
+  Biologically the unmatched synapses would die out of hunger, in the current 
+  implementation we increase their improtance in hope that (a few) will start 
+  to act on a subset of the input and become a new feature. 
```

Comment:
"If a column's connected synapses do not overlap well with any inputs often enough (as measured by `OverlapDutyCycles`), its permanence values are boosted."